k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -97,17 +113,19 @@ class EndpointsNamer(
     }
 
     def get(name: String): NsCache = synchronized {
-      caches.get(name) match {
+      caches.sample.get(name) match {
         case Some(ns) => ns
         case None =>
           val ns = new NsCache(name)
           val closable = retryToActivity { watch(name, ns) }
           _watches += (name -> closable)
-          caches += (name -> ns)
+          caches() = caches.sample + (name -> ns)
```

Comment:
We may want to note (in a comment) that `caches` must be updated with `synchronized` -- otherwise this could be potentially unsafe.